crates/deltalake-core/src/operations/write.rs

```diff
@@ -415,12 +476,31 @@ impl std::future::IntoFuture for WriteBuilder {
                 this.write_batch_size,
                 this.writer_properties,
                 this.safe_cast,
+                this.overwrite_schema,
             )
             .await?;
             actions.extend(add_actions.into_iter().map(Action::Add));
 
             // Collect remove actions if we are overwriting the table
             if matches!(this.mode, SaveMode::Overwrite) {
+                // Update metadata with new schema
+                let table_schema = this
+                    .snapshot
+                    .physical_arrow_schema(this.log_store.object_store().clone())
+                    .await
+                    .or_else(|_| this.snapshot.arrow_schema())
+                    .unwrap_or(schema.clone());
+
+                if schema != table_schema {
+                    let mut metadata = this
+                        .snapshot
+                        .current_metadata()
+                        .ok_or(DeltaTableError::NoMetadata)?
+                        .clone();
+                    metadata.schema = schema.clone().try_into().unwrap();
+                    let metadata_action = Metadata::try_from(metadata).unwrap();
+                    actions.push(Action::Metadata(metadata_action));
```

Comment:
let's handle these unwraps.

Reply:
@roeap Can you check if the change is ok? I think you want that it returns the error, so I used the question mark operator

Reply:
yes, exactly right :)